cli/lsp/completions.rs

```diff
@@ -682,6 +800,232 @@ mod tests {
     );
   }
 
+  #[tokio::test]
+  async fn test_get_npm_completions() {
```

Comment:
The tests here are very high level. For example, it's creating an http server, temp dir, http client, progress bars, npm api, cache. We can slim this down a lot for unit testing purposes and then maybe add a single basic integration test in `tests/integration/lsp_tests.rs`. The registry response is also far away from these tests in test_util/src/lib.rs which makes it harder to follow.

We could improve this by making a seam for the npm api like so:

```
#[async_trait::async_trait]
pub trait NpmSearchApi {
  async fn search(
    &self,
    query: &str,
  ) -> Result<Arc<Vec<String>>, AnyError>;
}

async fn get_npm_completions(
  specifier: &str,
  range: &lsp::Range,
  npm_api: &impl NpmSearchApi,
) -> Option<Vec<lsp::CompletionItem>> {
```

Then in the tests we can inject responses for different queries:

```rs
struct FakeNpmSearchApi(HashMap<String, Result<Arc<Vec<String>>, String>>);

#[async_trait::async_trait]
impl NpmSearchApi for FakeNpmSearchApi {
  async fn search(
    &self,
    query: &str,
  ) -> Result<Arc<Vec<String>>, AnyError> {
    self.0.get(query) // etc... with actual code that works here
  }
}
```

Then we could implement this trait for whatever struct we want in the real CLI which could then be used for a basic integration test to ensure all the plumbing is hooked up correctly or anything else that should be tested at an integration level.

This would allow us to keep everything close together and wouldn't require tying the lsp setup code to these tests.

Reply:
Ended up with:
```rust
async fn get_npm_completions(
  specifier: &str,
  range: &lsp::Range,
  npm_registry_api: &impl NpmRegistryApi,
  npm_search_api: &impl NpmSearchApi,
) -> Option<Vec<lsp::CompletionItem>> {}
```
since search and package metadata are separate. Also it's still using `CliNpmRegistryApi` as the implementer for `NpmSearchApi` for convenience